Deprioritize time formats when using local_date

alesolano opened the pull request:
Rationale:
If we are working with dates and symbol formats (e.g `local_date my_date :db`) we always get a time format and not a date format, which is annoying 😑. This is because the code [here](https://github.com/basecamp/local_time/blob/main/app/helpers/local_time_helper.rb#L49) prioritizes Time::DATE_FORMATS over Date::DATE_FORMATS when they share the same name. In other words, as `:db` is a builtin format in both Time and Dat, `local_date my_date :db` ends up adding hours, minutes and seconds when we only want the year, month and date.

Related: https://github.com/basecamp/local_time/issues/92

Here I'm basically splitting the `find_time_format` in two (`find_time_format` and `find_date_format`) and calling them accordingly.

This is just a draft to see whether this makes sense or not. My Ruby skills are not the best and I may struggle a lot creating the tests, but I'm willing to give this a try if you find this useful 😄.

josefarias commented:
Thanks for putting this together. I see where you're coming from. But I'm worried about breaking backwards compatibility for folks that depend on the original search order.

I think in most cases this could be worked around by using a different name for the date format:

```ruby
# config/initializers/date_and_time_formats.rb
Date::DATE_FORMATS[:iso] = "%Y-%m-%d"

# somewhere else...
local_date Date.today, :iso
```

I'm happy to keep discussing in https://github.com/basecamp/local_time/issues/92 if you have a case where this workaround isn't acceptable – although, granted, it's a tad inconvenient. Closing for now.

> My Ruby skills are not the best and I may struggle a lot creating the tests, but I'm willing to give this a try if you find this useful 😄.

Keep going! Reading and writing code like this is a great way to improve.

alesolano commented:
> I'm worried about breaking backwards compatibility for folks that depend on the original search order.

Thanks @josefarias, that makes total sense! The workaround, though not ideal, does the trick. Anyway, maybe we can consider this for version 3.0 😄

josefarias commented:
@alesolano having thought about this some more – it's weird that this library would require you to rename custom formats.

`DATE_FORMATS` are a [rails](https://github.com/rails/rails/blob/95af5fce71cef4492d3ebf3bae6c2b5748f2d9e1/activesupport/lib/active_support/core_ext/date/conversions.rb#L9-L21) [feature](https://github.com/rails/rails/blob/95af5fce71cef4492d3ebf3bae6c2b5748f2d9e1/activesupport/lib/active_support/core_ext/time/conversions.rb#L8-L26). If we're going to build on that, then we should follow what rails does. And it does know which formats to look for:

```ruby
>> Date.today.to_formatted_s(:db)
=> "2023-06-23"
>> Time.now.to_formatted_s(:db)
=> "2023-06-23 09:55:54"
```

Further, if people are rendering `Time` objects using `local_date`, or defining date formats inside `Time::DATE_FORMATS` because of this library's search order — then that's just clearly building on top of a bug. I think it might be reasonable to fix this.

We can reconsider the decision in my previous comment once we have some code to look at. If you're up for it, I'll leave some comments for your consideration on this PR. Just know you'll need to rebase or start from scratch since the code you're touching recently changed in https://github.com/basecamp/local_time/pull/135.

Please do let me know if you plan to continue working on this. Thanks!

josefarias commented on app/helpers/local_time_helper.rb:
```diff
@@ -12,9 +12,15 @@ def local_time(time, options = nil)
+    options[:data] ||= {}
+    options[:data].merge! local: :time, type: :date, format: format
+
+    time_tag time, time.strftime(format), options
```
I think we can still call `local_time` here if we calculate the correct format in `#local_date` and pass it as a string (as opposed to a symbol) to `local_time`.

josefarias commented on app/helpers/local_time_helper.rb:
```diff
@@ -42,20 +48,38 @@ def utc_time(time_or_date)
       end
     end
 
+    def find_i18n_time_format(format); find_i18n_format(format, :time); end
```
Let's use multi-line method definitions, since we don't do single-line anywhere else in this codebase

josefarias commented on app/helpers/local_time_helper.rb:
```diff
@@ -42,20 +48,38 @@ def utc_time(time_or_date)
+      I18n.t("#{type}.formats.#{format}", default: "").presence
+    end
+
+    def find_ruby_time_format(format); Time::DATE_FORMATS[format]; end
+    def find_ruby_date_format(format); Date::DATE_FORMATS[format]; end
```
Now that we've pulled the ruby format lookup into its own method, I think the discernment can go there. Maybe something like:

```ruby
def ruby_time_or_date_format(name, prefer: :time)
  if prefer == :time
    Time::DATE_FORMATS.with_indifferent_access[name] || Date::DATE_FORMATS.with_indifferent_access[name]
  else
    Date::DATE_FORMATS.with_indifferent_access[name] || Time::DATE_FORMATS.with_indifferent_access[name]
  end
end
```

Same goes for i18n lookup.

Feel free to go with something else if it makes sense, though.

josefarias replied:
Also, the name `ruby_time_or_date_format` was chosen to reflect that `time` has preference over `date` – thus coming first in the method name. This would now be communicated via the named param. So renaming might be a good idea. Just make sure to pick something that won't clash with other helpers in people's apps.

josefarias reviewed (commented):
We'll also want some tests, for sure. You can take a look at `test/helpers/local_time_helper_test.rb` for inspiration and run them with `rake test:helpers` or `rake test` if you also want to run the JS tests. Thanks!

josefarias commented:
@alesolano Hey! How are things going on this? I might work on a new release in the coming weeks and would be great to include this.

alesolano commented:
> @alesolano Hey! How are things going on this? I might work on a new release in the coming weeks and would be great to include this.

Oh, that's great. Sure, I can put some time into this this Friday and the following one if necessary. After this Friday I'll give you an estimation about whether I think I can get it done by next week. Thanks for the ping! @josefarias

alesolano commented:
FYI: Not my intention to close this. I just discarded my commit to sync my fork with the new changes. Hopefully this will reopen once I push my new commits.

alesolano commented:
Update: I think we can get it done by this week or next. I'll work a bit more on this and I'll request a review during the following days.

alesolano commented:
I'm happy with the code right now. I've changed maybe more things than I should, so don't be afraid to request to change it back to something more similar to the original code. Also, feel free to take over this and update it yourself.

Main things to discuss:
- I added two more tests including an ambiguous format that exists both for time and date. Here, using local_time and local_date should output different results. Previously, it would show the same: a time format.
- An existing test fails `test_local_date_with_missing_i18n_and_ruby_format` because it expects that the default format for a local date is a time format. Should we change this?
- Prioritizing date formats over time formats involves duplicating almost every method. I've changed them instead to almost all of them accept a "type" argument.
- Format24 does not make sense for date formats. I changed the private methods to return hashes instead of arrays with "nil" as second element. I find this cleaner to make the format24 coexist with dates. Happy to change it if we don't agree.

josefarias commented:
@alesolano thanks for this. This is good work!

I do think there's potential for simplification.

> feel free to take over this and update it yourself.

I'll take you up on that – but I want to be clear if we do end up merging this I'll make sure you get proper credit.

Here's where I ended up: https://github.com/basecamp/local_time/pull/143. Let me know what you think.

> I might work on a new release in the coming weeks and would be great to include this.

I ran out of time for this, so will probably take longer to cut a release myself.

josefarias commented on app/helpers/local_time_helper.rb:
```diff
@@ -54,51 +60,65 @@ def extract_options_and_value(options, value_key = nil)
+    def find_ruby_formats(name, type)
+      format_data = ruby_format(name, type)
+      if format_data.is_a?(Proc)
+        { format: default_format(type) }
+      elsif type == :time
+        { format: format_data, format24: ruby_format("#{name}_24h") }
+      else
+        { format: format_data }
```
Returning formats as a hash is a nice idea. But I'd prefer if the method didn't have to know whether it needs to return two formats or just one. And since we've already gone with returning a "tuple" (array, really) I'd rather we stick with that.

josefarias commented on test/helpers/local_time_helper_test.rb:
```diff
@@ -134,7 +141,14 @@ def test_local_date_with_ruby_format
+  end
+
   def test_local_date_with_missing_i18n_and_ruby_format
+    # This one fails because we expect the default time format and we get the default date format.
+    # Maybe we should change this test?
```
Yup! Let's change

josefarias reviewed (commented):
Already left a comment above sharing an alternative PR, but thought I'd comment my thoughts anyway.

josefarias commented on app/helpers/local_time_helper.rb:
```diff
@@ -2,19 +2,25 @@ module LocalTimeHelper
+    options[:data][:local] = :time
+    options[:data].merge! find_format_data(format, :date)
+
+    time_tag time, time.strftime(options[:data][:format]), options
```
`#local_date` is _really_ just `#local_time` called with a date format. So it'd be great if we called `local_time` from here instead of `time_tag`. We should be able to do that if we determine the format before reaching this point and then passing it as a string to `#local_time`.

alesolano commented:
Thanks a lot @josefarias!! Sorry for not being able to find time to contribute more. Very, very happy for this :)

josefarias commented:
> Sorry for not being able to find time to contribute more

Don't be! You got this off to a great start. Hope to see more contributions from you in the future, here or elsewhere.

Thanks again!
